### PR TITLE
add bubble_sort_by function with yield test

### PR DESCRIPTION
**Pull request by janessatran**
Working through this exercise on the Odin project: [Project 1: Bubble Sort - Assignment 1](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks?ref=lnav) 

I don't really understand yet why we are using the yield block in our bubble_sort_by function. Is it just to practice using yield? From what I understand so far, yield allows you to override a function with new functionality placed in a block. I suppose I just don't understand why that is useful in this context.

### CODE REVIEW

File: bubble_sort.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        swapped = true
+      end 
+    end 
+  end
+  arr
+end 
+
+
+def bubble_sort_by(arr)
+  if arr.length == 1
```
Comment (slabounty) [inline bubble_sort.rb:20]: `return arr if arr.length == 1`

File: bubble_sort.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      if comparison > 0
+        arr[idx], arr[idx+1] = arr[idx+1], arr[idx]
+        swapped = true
+      end 
+    end 
+  end
+  arr
+end 
+
+bubble_sort_by(['hello', 'hi', 'hey']) do |left, right|
```
Comment (slabounty) [inline bubble_sort.rb:37]: So it looks like you have a test(ish) sort of thing here. Fine for looking at, but generally we'd remove in real life.

File: bubble_sort.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+def bubble_sort_by(arr)
+  if arr.length == 1
+    return arr
+  end
+  swapped = true
+  while swapped
+    swapped = false
+    (0..arr.length-2).each do |idx|
+      comparison = yield arr[idx], arr[idx+1]
```
Comment (slabounty) [inline bubble_sort.rb:27]: So ... what this is doing is allowing you to sort however you like. In the "test" below, you've chosen to sort by the length of the string you're passing in. In general, with something like this a negative means left is "less than" right, 0 means equal, and positive means left is "greater than" right. You can use the technique here for interesting sorting methodologies. This is a good exercise, but the ruby enumerable (used by arrays etc.) you can do a similar thing.

File: spec/bubble_sort_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+describe '#bubble_sort' do
+
+  it 'compares each element in an array, swaps them, until the array is sorted' do
+    expect((bubble_sort([4,5,2,1,6]))).to eq([1,2,4,5,6])
+  end
+
+  it 'returns the array if the array size is 1' do
+    expect((bubble_sort([1]))).to eq([1])
+  end
+
```
Comment (slabounty) [inline spec/bubble_sort_spec.rb:11]: Extra blank line.

File: spec/bubble_sort_spec.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+require ('./bubble_sort')
+
+describe '#bubble_sort' do
+
```
Comment (slabounty) [inline spec/bubble_sort_spec.rb:4]: Extra blank line.

File: spec/bubble_sort_spec.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+describe '#bubble_sort_by' do
+
+  it 'uses the yield function' do
+    result = bubble_sort_by(['hello', 'hi', 'hey']) do |left, right|
+      left.length - right.length
+    end
+    expect(result).to eq(['hi', 'hey', 'hello'])
+  end 
+
```
Comment (slabounty) [inline spec/bubble_sort_spec.rb:25]: Extra blank line.

File: bubble_sort.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+def bubble_sort(arr)
+  if arr.length == 1
+    return arr
+  end
+  swapped = true
+  while swapped
+    swapped = false
+    (0..arr.length-2).each do |idx|
+      if arr[idx] > arr[idx+1]
```
Comment (slabounty) [inline bubble_sort.rb:9]: This is relying on the "native" comparison for whatever you pass in. It wouldn't probably work (or at least not how you expect) if you passed in an array of some objects.

File: bubble_sort.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        arr[idx], arr[idx+1] = arr[idx+1], arr[idx]
+        swapped = true
+      end 
+    end 
+  end
+  arr
+end 
+
+bubble_sort_by(['hello', 'hi', 'hey']) do |left, right|
+  left.length - right.length
```
Comment (slabounty) [inline bubble_sort.rb:38]: If you'd passed in an array of some objects, this'd be where you could say use a method on those objects for comparison.

**Review by slabounty — CHANGES_REQUESTED** [review]
Mostly explanations. A couple of minor suggestions.

File: bubble_sort.rb (hunk 3 above)
Comment (slabounty) [reply bubble_sort.rb:27]: So ... had a few minutes this morning and came up with this ...
```ruby
janessa = OpenStruct.new(age: 24, height: 67)
jennifer = OpenStruct.new(age: 24, height: 60)
scott = OpenStruct.new(age: 29, height: 66)

sort_by_age = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
  left.age - right.age
end
puts "sort_by_age = #{sort_by_age}"

sort_by_height = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
  left.height - right.height
end
puts "sort_by_height = #{sort_by_height}"
```

which if you run in a console with your code you get ...

```
[19] pry(main)> janessa = OpenStruct.new(age: 24, height: 67)
=> #<OpenStruct age=24, height=67>
[20] pry(main)> jennifer = OpenStruct.new(age: 24, height: 60)
=> #<OpenStruct age=24, height=60>
[21] pry(main)> scott = OpenStruct.new(age: 29, height: 66)
=> #<OpenStruct age=29, height=66>
[22] pry(main)>
[23] pry(main)> sort_by_age = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
[23] pry(main)*   left.age - right.age
[23] pry(main)* end
=> [#<OpenStruct age=24, height=67>, #<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>]
[24] pry(main)> puts "sort_by_age = #{sort_by_age}"
sort_by_age = [#<OpenStruct age=24, height=67>, #<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>]
=> nil
[25] pry(main)>
[26] pry(main)> sort_by_height = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
[26] pry(main)*   left.height - right.height
[26] pry(main)* end
=> [#<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>, #<OpenStruct age=24, height=67>]
[27] pry(main)> puts "sort_by_height = #{sort_by_height}"
sort_by_height = [#<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>, #<OpenStruct age=24, height=67>]
=> nil
```

So, in the first case, we sort by the age of the people in the `OpenStruct` and in the 2nd case we sort by height.

**Review by slabounty — COMMENTED** [review]
Added an example as to why your ` bubble_sort_by` might be useful.

File: bubble_sort.rb (hunk 7 above)
Comment (janessatran) [reply bubble_sort.rb:9]: ooh good to know

File: bubble_sort.rb (hunk 8 above)
Comment (janessatran) [reply bubble_sort.rb:38]: ah, i see

**Pull request merged**